[ConstraintSystem] Delay constraint generation for single-statement closure body

xedin opened the pull request:
Attempt to infer a closure type based on its parameters and body
and put it aside until contextual type becomes available or it
has been determined that there is no context.

Once all appropriate conditions are met, generate constraints for
the body of the closure and bind it the previously inferred type.

Doing so makes it possible to pass single-statement closures as
an argument to a function builder parameter.

Resolves: SR-11628
Resolves: rdar://problem/56340587

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Augment `getCalleeLocator` with `getType` callback

This is necessary for ambiguity diagnostics because none of the
solutions are applied back to the constraint system which means
that `getCalleeLocator` needs an ability to query information
from different sources e.g. constraint system and/or solution.

xedin committed:
[ConstraintSystem] Add a separate constraint generation for closure body

Separate closure from top-level constraint generation because
its body shouldn't be considered when it comes to parent/weight
computation, otherwise top expression in body is going to get
disconnected from the closure.

xedin committed:
[ConstraintSystem] Introduce a mapping from closures to their inferred types

In preparation to change type-checking behavior of closures
we need to introduce a special mapping from closure expressions
to their inferred types (based on parameters/result and body).

xedin committed:
[ConstraintSystem] Introduce `resolveClosure` method to generate constraints for closure body

In preparation to change the way closures are handled in constraint system
let's introduce a special method which is responsible for generation of
constraints for closure body and connecting it to the rest of the
constraint system when contextual type becomes available.

xedin committed:
[ConstraintSystem] Introduce new `DefaultClosureType` constraint

This constraint connects type variable representing a closure
expression to its inferred type and behaves just like regular
`Defaultable` constraint expect for type inference where it's
used only if there are no contextual bindings available.

xedin committed:
[ConstraintSystem] Allow single-statement closures be used with function builders

While resolving closure use contextual locator information to
determine whether given contextual type comes from a "function builder"
parameter and if so, use special `applyFunctionBuilder` to open
closure body instead of regular constraint generation.

xedin committed:
[AST] NFC: Remove obsolete function builder type storage from `ParameterListInfo`

xedin committed:
[ConstraintSystem] Clarify `DefaultClosureType` constraint and add its own `simplify` method

xedin committed:
[ConstraintSystem] Add accessor to type var to determine if it represents a closure type

xedin committed:
[ConstraintSystem] Account for unapplied curried self while trying to retrieve function builder type for parameter

xedin committed:
[CSBindings] Prefer closure type binding over disjunction

Even if contextual closure type has type variables inside
prefer it over disjunction because it provides a lot of
contextual information early which helps to solve the system
faster.

xedin committed:
[ConstraintSystem] Delay constraint generation from single-statement closure body

Attempt to infer a closure type based on its parameters and body
and put it aside until contextual type becomes available or it
has been determined that there is no context.

Once all appropriate conditions are met, generate constraints for
the body of the closure and bind it the previously inferred type.

Doing so makes it possible to pass single-statement closures as
an argument to a function builder parameter.

Resolves: SR-11628
Resolves: rdar://problem/56340587

xedin committed:
[Diagnostics] Pinpoint contextual error location if it's related to a call to closure

For example `let _: B = { A() }()` should point to a closure result
instead of a call.

xedin committed:
[CSFix] Account for special closure handling in invalid trailing closure fix

Since closures don't get the type assigned right away anymore we
have to fetch it before any of the internal type variables could
be marked as holes.

xedin committed:
[CSGen] NFC: Remove obsolete closure handling logic

Since we no longer walk into closures it's possible to remove
`{enter, exit}Closure` and `getClosureParams`.

xedin committed:
[TypeChecker] NFC: Mark perf test-case for rdar://problem/20959612 as "fast"

This test-case has been improved by changes in closure expression
handling in constraint system, because the bodies are going to be
opened in order, the nesting yields linear performance.

xedin committed:
[Diagnostics] NFC: Adjust test-cases improved by changes to closure expression handling

xedin committed:
[TypeChecker] NFC: Mark test-case for rdar://problem/22022980 as "slow"

It used to be "fast" before because `selectDisjunction` used to pick
inner `==`s before outer one (which compares arrays) but now, since
constraint generation for the closure body is delayed, it looks
like the opposite is happening which makes this test-case go exponential.

xedin commented:
@swift-ci Please Build Toolchain macOS Platform

xedin committed:
[ConstraintSystem] Add more information retrieval callbacks to `getCalleeLocator`

`getCalleeLocator` has to be able to:

- Retrieve type for given expression via `getType`;
- Simplify given type via `simplifyType`; and
- Retrieve a selected overload choice for a given locator via `getOverloadFor`.

xedin commented:
@swift-ci Please Build Toolchain macOS Platform

xedin commented:
@swift-ci Please Build Toolchain macOS Platform

xedin commented on lib/Sema/ConstraintSystem.h:
```diff
@@ -2181,8 +2212,28 @@ class ConstraintSystem {
+      llvm::function_ref<Type(const Expr *)> getType,
+      llvm::function_ref<Type(Type)> simplifyType,
+      llvm::function_ref<Optional<SelectedOverload>(ConstraintLocator *)>
+          getOverloadFor);
```
@hamishknight Please take a look at the changes for `getCalleeLocator` there is now one in `ConstraintSystem` and one in the `Solution` because ambiguity diagnostics have to use unapplied solution to retrieve type/overload information.

hamishknight reviewed (approved):
`getCalleeLocator` changes look good to me!

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

CodaFi commented on lib/Sema/CSGen.cpp:
```diff
@@ -2050,38 +2036,79 @@ namespace {
+          Type externalType;
+          if (param->getTypeRepr()) {
+            auto declaredTy =
+                closure->mapTypeIntoContext(param->getInterfaceType());
```
You should just be able to say `param->getType()`.

xedin replied:
Ah, thank you!

CodaFi commented on lib/Sema/CSGen.cpp:
```diff
@@ -3643,11 +3648,12 @@ namespace {
         if (closure->hasSingleExpressionBody()) {
-          CG.enterClosure(closure);
+          auto closureType = CG.visitClosureExpr(closure);
+          if (!closureType)
```
This should never be `Type()`

xedin replied:
It can be in `inferClosureType` returns type with an error...

CodaFi commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -6354,18 +6360,38 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
+    TypeMatchOptions flags, ConstraintLocatorBuilder locator) {
+  closureType = getFixedTypeRecursive(closureType, flags, /*wantRValue=*/true);
+
+  if (closureType->isTypeVariableOrMember()) {
```
I wonder if it makes sense to add `hasTypeVariable()` here.

xedin replied:
I don't think so because we really want to keep this constraint until the closure is resolved and the type it's bound to can (and probably would) have type variables. This constraint serves dual purpose - source of type for the closure when there is not contextual information available e.g. `_ = { $0 + 1 }` and to connect closure type variable to its inferred parameters and result type + any outer parameters used in the body until the closure is resolved. When closure type variable becomes bound to inferred type (closure is "resolved") everything becomes connected, so `DefaultClosureType` is obsolete from that moment on. I was thinking instead of `DefaultCloureType` to go with `InferredClosureType`, maybe that's a better name...

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test compiler performance

xedin commented:
@swift-ci please test compiler performance

xedin commented:
@swift-ci please test source compatibility debug

xedin commented:
@swift-ci please test source compatibility debug

xedin commented:
@swift-ci please test source compatibility release

xedin committed:
[CSGen] Use `getType` when refering to closure parameters in expression context

xedin commented:
I'm not sure why but some of the builds where UPASS which I think contributed to the difference in number of constraint scopes, but I think that's okay because it didn't contribute to build time...

xedin committed:
[ConstraintSystem] Switch to pair-wise type binding differentiation for ranking

Instead of trying to hold a "global" set of type variable differences
let's use pair-wise comparison instead because in presence of generic
overloads such would be more precise.

If there are N solutions for a single generic overload we currently
relied on "local" comparison to detect the difference, but it's not
always possible to split the system to do one. Which means higher
level comparisons have to account for "local" (per overload choice)
differences as well otherwise ranking would loose precision.

xedin committed:
[TypeChecker] Use special locator while matching function builder body to closure result type

xedin committed:
[CSRanking] Account for that fact that some bindings can be type holes

xedin committed:
[ConstraintSystem] Add pair-wise differentiation of key path dynamic member overloads

Single type of keypath dynamic member lookup could refer to different
member overlaods, we have to do a pair-wise comparison in such cases
otherwise ranking would miss some viable information e.g.
`_ = arr[0..<3]` could refer to subscript through writable or read-only
key path and each of them could also pick overload which returns `Slice<T>`
or `ArraySlice<T>` (assuming that `arr` is something like `Box<[Int]>`).

xedin committed:
[TypeChecker] NFC: Adjust a couple of test-case affected by closure constraint generation changes

xedin commented:
@DougGregor @hborla This is ready for a final review. I'm going to run CI meanwhile.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please smoke test compiler performance

xedin committed:
[ConstraintSystem] Attempt types linked via subtyping in reverse order (supertypes first)

If there is a subtype relationship between N types we have to make
sure that type chain is attempted in reverse order because we can't
infer bindings transitively through type variables and attempting
it in any other way would miss potential bindings across the chain.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please smoke test compiler performance

DougGregor commented on lib/Sema/CSGen.cpp:
```diff
@@ -3733,34 +3738,11 @@ namespace {
 
       if (auto closure = dyn_cast<ClosureExpr>(expr)) {
         if (closure->hasSingleExpressionBody()) {
-          CG.exitClosure(closure);
```
Do we need the single-expression closure special case at all?

xedin replied:
Ah you are right, we do not need that, both single- and multi-statement closures should behave the same.

DougGregor reviewed (approved):
This is awesome!

xedin commented:
@swift-ci please smoke test

xedin committed:
[CSGen] Unify handling of single- and multi-statement closures

Both multi- and single-statement closures now behave the same way
when it comes to constraint generation, because the body is opened
only once contextual type becomes available or it's impossible
to find one e.g. `_ = { 1 }`.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
Alright, here we go :)
